pixeltable/type_system.py

```diff
@@ -927,149 +941,116 @@ def supertype(self, other: ColumnType) -> ArrayType | None:
-        return f'Array[{self.shape}, {self.pxt_dtype}]'
+        return f'Array[{self.shape}, {self.dtype.name}]'
 
     @classmethod
     def _from_dict(cls, d: dict) -> ColumnType:
+        assert 'numpy_dtype' in d
+        dtype = None if d['numpy_dtype'] is None else np.dtype(d['numpy_dtype'])
         assert 'shape' in d
-        assert 'dtype' in d
         shape = None if d['shape'] is None else tuple(d['shape'])
-        dtype = None if d['dtype'] is None else cls.make_type(cls.Type(d['dtype']))
         return cls(shape, dtype, nullable=d['nullable'])
 
-    @classmethod
-    def from_np_dtype(cls, dtype: np.dtype, nullable: bool) -> ColumnType | None:
-        """
-        Return pixeltable type corresponding to a given simple numpy dtype
-        """
-        if np.issubdtype(dtype, np.integer):
-            return IntType(nullable=nullable)
-
-        if np.issubdtype(dtype, np.floating):
-            return FloatType(nullable=nullable)
-
-        if dtype == np.bool_:
-            return BoolType(nullable=nullable)
-
-        if np.issubdtype(dtype, np.str_):
-            return StringType(nullable=nullable)
-
-        if np.issubdtype(dtype, np.character):
-            return StringType(nullable=nullable)
-
-        if np.issubdtype(dtype, np.datetime64):
-            unit, _ = np.datetime_data(dtype)
-            if unit in ['D', 'M', 'Y']:
-                return DateType(nullable=nullable)
-            else:
-                return TimestampType(nullable=nullable)
-
-        return None
-
     @classmethod
     def from_literal(cls, val: np.ndarray, nullable: bool = False) -> ArrayType | None:
-        # determine our dtype
         assert isinstance(val, np.ndarray)
-        pxttype: ColumnType | None = cls.from_np_dtype(val.dtype, nullable)
-        if pxttype is None:
+        if val.dtype.type not in ARRAY_SUPPORTED_NUMPY_DTYPES:
             return None
-        return cls(val.shape, dtype=pxttype, nullable=nullable)
+        return cls(val.shape, dtype=val.dtype, nullable=nullable)
 
-    def is_valid_literal(self, val: np.ndarray) -> bool:
+    def _to_json_schema(self) -> dict[str, Any]:
+        schema: dict[str, Any] = {'type': 'array'}
+        if self.dtype == np.str_:
+            schema.update({'items': {'type': 'str'}})
+        elif self.dtype is not None:
+            schema.update({'items': {'type': str(self.dtype)}})
+        return schema
+
+    def _validate_literal(self, val: Any) -> None:
         if not isinstance(val, np.ndarray):
-            return False
+            raise TypeError(f'Expected numpy.ndarray, got {val.__class__.__name__}')
 
-        # If a dtype is specified, check that there's a match
-        if self.dtype is not None and not np.issubdtype(val.dtype, self.numpy_dtype()):
-            return False
+        # If column type has a dtype, check for compatibility. Note: the exact dtype match is not required. We allow
+        # literals whose dtypes can be cast to the column's.
```

Comment:
See comment above regarding cast. I think we need to guarantee the physical format, not just whether data conversion is possible.

Reply:
PTAL